storage/mysql/mysql.go

```diff
@@ -0,0 +1,669 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/openfga/openfga/pkg/id"
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/telemetry"
+	tupleUtils "github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/storage"
+	"github.com/pkg/errors"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	defaultMaxTuplesInWrite     = 100
+	defaultMaxTypesInDefinition = 100
+)
+
+type MySQL struct {
+	db                       *sql.DB
+	tracer                   trace.Tracer
+	logger                   logger.Logger
+	maxTuplesInWrite         int
+	maxTypesInTypeDefinition int
+}
+
+type MySQLOption func(*MySQL)
+
+func WithMaxTuplesInWrite(maxTuples int) MySQLOption {
+	return func(m *MySQL) {
+		m.maxTuplesInWrite = maxTuples
+	}
+}
+
+func WithMaxTypesInTypeDefinition(maxTypes int) MySQLOption {
+	return func(m *MySQL) {
+		m.maxTypesInTypeDefinition = maxTypes
+	}
+}
+
+func WithLogger(l logger.Logger) MySQLOption {
+	return func(m *MySQL) {
+		m.logger = l
+	}
+}
+
+func WithTracer(t trace.Tracer) MySQLOption {
+	return func(m *MySQL) {
+		m.tracer = t
+	}
+}
+
+func NewMySQLDatastore(uri string, opts ...MySQLOption) (*MySQL, error) {
+	m := &MySQL{}
+
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	if m.logger == nil {
+		m.logger = logger.NewNoopLogger()
+	}
+
+	if m.tracer == nil {
+		m.tracer = telemetry.NewNoopTracer()
+	}
+
+	if m.maxTuplesInWrite == 0 {
+		m.maxTuplesInWrite = defaultMaxTuplesInWrite
+	}
+
+	if m.maxTypesInTypeDefinition == 0 {
+		m.maxTypesInTypeDefinition = defaultMaxTypesInDefinition
+	}
+
+    policy := backoff.NewExponentialBackOff()
+	policy.MaxElapsedTime = 1 * time.Minute
+	var db *sql.DB
+	attempt := 1
+	err := backoff.Retry(func() error {
+		var err error
+		db, err = sql.Open("mysql", uri)
```

Comment:
From the docs: _Open may just validate its arguments without creating a connection to the database. To verify that the data source name is valid, call Ping._